dev/error_message_refiner.py

```diff
@@ -0,0 +1,235 @@
+                content,
+                re.DOTALL,
+            )
+            if function_content and search_string in function_content.group(0):
+                return function_content.group(0)
+
+    return None
+
+
+def _get_error_function(error_class: str) -> str:
+    """
+    Retrieves the name of the error function that triggers the given error class.
+    """
+    search_string = error_class
+    matched_file = _git_grep_files(search_string)
+    err_func = _find_function(matched_file, search_string)
+    return err_func
+
+
+def _get_source_code(error_class: str) -> str:
+    """
+    Retrieves the source code of a function where the given error class is being invoked.
+    """
+    search_string = error_class
+    matched_file = _git_grep_files(search_string)
+    err_func = _find_function(matched_file, search_string)
+    source_file = _git_grep_files(err_func, exclude=matched_file)
+    source_code = _find_func_body(source_file, err_func)
+    return source_code
+
+
+def _get_error_message(error_class: str) -> str:
+    """
+    Returns the error message from the provided error class
+    listed in core/src/main/resources/error/error-classes.json.
+    """
+    with open(PATH_TO_ERROR_CLASS) as f:
+        error_classes = json.load(f)
+
+    error_message = error_classes.get(error_class)
+    if error_message:
+        return error_message["message"]
+    else:
+        return f"Error message not found for class: {error_class}"
+
+
+def ask_chat_gpt(error_class: str, gpt_version: str) -> Tuple[str, str]:
+    """
+    Requests error message improvement from Chat GPT.
+    Returns a tuple containing the old error message and the refined error message.
+    """
+    old_error_message = " ".join(_get_error_message(error_class))
+    error_function = _get_error_function(error_class)
+    source_code = _get_source_code(error_class)
+    message = (
+        "I want to improve the error message.\n"
+        "The error message should be user-friendly, "
+        "providing a clear description of the error's cause, "
+        "as well as suggesting the best possible actions for the user to take.\n"
+        f'Please improve the following error message: "{old_error_message}"\n'
+        "PLEASE NOTE that the angle brackets in the error message represent the placeholder "
+        "for the error message parameter, so they should not be modified or removed."
+        f"For more detail, the error message is triggered through a error function "
+        f'named "{error_function}", and the source code that actually calls the error function '
+        f"is as follows:\n\n{source_code}\n\n"
+        "When improving the error, please also refer to the source code provided above "
+        "to provide the user with more useful information.\n"
+        "Also, please only show improved error message without any additional explanation "
+        "so that I can just copy and paste your answer.\n"
```

Comment:
https://spark.apache.org/error-message-guidelines.html is the guideline. Let us add them to the prompt.

Reply:
Just added the link to the prompt. Thanks!

Reply:
Updated the prompt to respect the [Error Message Guidelines](https://spark.apache.org/error-message-guidelines.html), and also polished the miscellaneous things. FYI @gatorsmile